test/benchmarks/material/button-harness/testbed.perf-spec.ts

```diff
@@ -0,0 +1,135 @@
+    await benchmark('(baseline) click first button', async () => {
+      const button = getButtonWithText(FIRST_BUTTON);
+      button.click();
+    });
+  });
+
+  it('(baseline) should click the middle button', async () => {
+    await benchmark('(baseline) click middle button', async () => {
+      const button = getButtonWithText(MIDDLE_BUTTON);
+      button.click();
+    });
+  });
+
+  it('(baseline) should click the last button', async () => {
+    await benchmark('(baseline) click last button', async () => {
+      const button = getButtonWithText(LAST_BUTTON);
+      button.click();
+    });
+  });
+
+  it('(baseline) should click all of the buttons', async () => {
+    await benchmark('(baseline) click every button', async () => {
+      const buttons = document.getElementsByTagName('button');
+      for (let i = 0; i < buttons.length; i++) {
+        const button = buttons[i];
+        button.click();
+      }
+    });
+  });
+});
+
+describe('performance tests for the testbed harness', () => {
+  let fixture: ComponentFixture<ButtonHarnessTest>;
+  let loader: HarnessLoader;
+
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      imports: [MatButtonModule],
+      declarations: [ButtonHarnessTest],
+    }).compileComponents();
+
+    fixture = TestBed.createComponent(ButtonHarnessTest);
+    fixture.detectChanges();
+    loader = TestbedHarnessEnvironment.loader(fixture);
+  });
+
+  it('should retrieve all of the buttons', async () => {
+    await benchmark('get every button', async () => {
+      await loader.getAllHarnesses(MatButtonHarness);
+    });
+  });
+
+  it('should click the first button', async () => {
+    await benchmark('click first button', async () => {
+      const button = await loader.getHarness(MatButtonHarness.with({text: FIRST_BUTTON}));
+      await button.click();
+    });
+  });
+
+  it('should click the middle button', async () => {
+    await benchmark('click middle button', async () => {
+      const button = await loader.getHarness(MatButtonHarness.with({text: MIDDLE_BUTTON}));
+      await button.click();
+    });
+  });
+
+  it('should click the last button', async () => {
+    await benchmark('click last button', async () => {
+      const button = await loader.getHarness(MatButtonHarness.with({text: LAST_BUTTON}));
+      await button.click();
+    });
+  });
+
+  it('should click all of the buttons', async () => {
+    await benchmark('click every button', async () => {
+      const buttons = await loader.getAllHarnesses(MatButtonHarness);
+      for (let i = 0; i < buttons.length; i++) {
+        const button = buttons[i];
+        await button.click();
+      }
+    });
+  });
+
+  // To see the benchmarks for this test, uncomment the it test below.
+  //
+  // I don't know how to force karma_web_test to show logs in the console so we are using this as
+  // a solution for now.
+  // it('should fail intentionally', () => expect(1).toBe(2));
```

Comment:
I think this is symptomatic that this should actually not be a test. We don't run any assertions. logic-wise, nor do we compare the durations to previous runs. At the same time though, I don't think there is a good way to run a non-test in browsers right now. Framework runs such benchmarks in Node using a mock DOM (not ideal either).

Can you elaborate a bit on when someone intends to run this benchmark? Is it a full manual thing? If so, we could make the tests always fail, right?

Reply:
I agree that ideally they wouldn't even be set up as tests, though for now I think its fine to have them as tests that only get run manually and always fail

Reply:
Yeah, failing always is what I had in mind too. It's obvious there is no better way to do this right now, unfortunately 😞